### PR TITLE
Feature: Add an option to select hashes to show

### LINKED ISSUE #11623: Feature: Add an option to select hash algorithms to be calculated.
### What feature or improvement do you think would benefit Files?

Since calculating hash values is a heavy process for large files, I think there should be an option to not calculate unnecessary hash values.

### Requirements

Add a menu flyout in the hash page for selecting hash algorithms to be calculated.

### Files Version

Dev

### Windows Version

10.0.22621.1265

### Comments

_No response_

### PR DESCRIPTION
**Pull request by hishitetsu**
**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Closes #11623 

**Notes**
* SHA384 and SHA512 is turned off by default.
* Also fixed the bug that the hash value is not calculated correctly.

**Validation**
How did you test these changes?
- [X] Built and ran the app
- [X] Tested the changes for accessibility

**Screenshots**
image

### CODE REVIEW

**Comment by onein528** [comment]
These 5 is default. more ones should be like this.

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 1]
```diff
@@ -43,74 +37,102 @@ public HashInfoItem SelectedItem
...
+
+							case "SHA384":
+								hashInfoItem.HashValue = await ChecksumHelpers.CreateSHA384(_stream, _cancellationTokenSource.Token);
+								break;
+
+							case "SHA512":
+								hashInfoItem.HashValue = await ChecksumHelpers.CreateSHA512(_stream, _cancellationTokenSource.Token);
+								break;
+						}
+						hashInfoItem.IsCalculated = true;
```
Comment (onein528) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:119]: You can use 'Switch Expression'

File: src/Files.App/Views/Properties/HashesPage.xaml [hunk 2]
```diff
@@ -79,14 +65,43 @@
...
+						</Button>
 					</Grid>
 				</ListView.Header>
 
 				<!--  Template  -->
 				<ListView.ItemTemplate>
 					<DataTemplate x:DataType="models:HashInfoItem">
-						<ListViewItem IsSelected="{x:Bind IsSelected, Mode=TwoWay}">
+						<ListViewItem IsSelected="{x:Bind IsSelected, Mode=TwoWay}" Visibility="{x:Bind IsEnabled, Mode=OneWay}">
 							<Grid ColumnSpacing="12">
```
Comment (onein528) [inline src/Files.App/Views/Properties/HashesPage.xaml:105]: - Use Converter 'BoolToVisibility' since IsEnabled must be boolean.
- Remove Mode=OneWay due to a warning WMC1506.

File: src/Files.App/Views/Properties/HashesPage.xaml [hunk 3]
```diff
@@ -106,14 +121,15 @@
 									Grid.Column="2"
 									VerticalAlignment="Center"
 									Style="{StaticResource BodyTextBlockStyle}"
-									Text="{x:Bind HashValue}"
+									Text="{x:Bind HashValue, Mode=OneWay}"
```
Comment (onein528) [inline src/Files.App/Views/Properties/HashesPage.xaml:124]: As above

File: src/Files.App/Views/Properties/HashesPage.xaml [hunk 4]
```diff
@@ -106,14 +121,15 @@
...
 									TextTrimming="CharacterEllipsis"
 									TextWrapping="NoWrap"
 									ToolTipService.ToolTip="{x:Bind HashValue}" />
 
 								<!--  Copy button  -->
 								<Button
 									x:Name="CopyHashButton"
+									x:Load="{x:Bind IsCalculated, Mode=OneWay}"
```
Comment (onein528) [inline src/Files.App/Views/Properties/HashesPage.xaml:132]: As above

File: src/Files.App/Views/Properties/HashesPage.xaml [hunk 5]
```diff
@@ -134,10 +150,10 @@
 			<!--  Unavailable Label  -->
 			<TextBlock
 				x:Name="UnavailableTextBlock"
-				Margin="16,56,16,16"
+				Margin="16,16,16,16"
```
Comment (onein528) [inline src/Files.App/Views/Properties/HashesPage.xaml:153]: Is this alright?

File: src/Files.App/Views/Properties/HashesPage.xaml.cs [hunk 6]
```diff
@@ -21,7 +21,10 @@ protected override void OnNavigatedTo(NavigationEventArgs e)
 		{
 			var np = (MainPropertiesPage.PropertyNavParam)e.Parameter;
 			if (np.navParameter is ListedItem listedItem)
-				HashesViewModel = new(listedItem);
+			{
+				HashesViewModel = new();
+				HashesViewModel.Initialize(listedItem);
```
Comment (onein528) [inline src/Files.App/Views/Properties/HashesPage.xaml.cs:26]: Please do this in constructor of the ViewModel

File: src/Files.App/Views/Properties/HashesPage.xaml.cs [hunk 7]
```diff
@@ -35,12 +38,17 @@ private void CopyHashButton_Click(object sender, RoutedEventArgs e)
 			Windows.ApplicationModel.DataTransfer.Clipboard.SetContent(dp);
 		}
 
-		private async void BasePropertiesPage_Loaded(object sender, RoutedEventArgs e)
+		private bool _Cancel;
```
Comment (onein528) [inline src/Files.App/Views/Properties/HashesPage.xaml.cs:41]: _cancel

File: src/Files.App/Views/Settings/AboutPage.xaml [hunk 8]
```diff
@@ -160,7 +160,7 @@
 							Margin="10"
 							Background="Transparent"
 							LinkClicked="MarkdownTextBlock_LinkClicked"
-							Text="{x:Bind ViewModel.ThirdPartyNotices, Mode=TwoWay, UpdateSourceTrigger=PropertyChanged}" />
+							Text="{x:Bind ViewModel.ThirdPartyNotices, Mode=OneWay}" />
```
Comment (onein528) [inline src/Files.App/Views/Settings/AboutPage.xaml:163]: Avoid unrelated changes.
FYI, When you create AboutPage improvements in another PR, please move methods in AboutPage's code-behind to the ViewModel.

**Review by onein528 — CHANGES_REQUESTED** [review]
Sort of. Otherwise, LGTM.

**Comment by hishitetsu** [comment]
A critical bug was found. `Stream` is not rewound, so the hash value is not calculated correctly.
Draft until it is fixed.

**Comment by hishitetsu** [comment]
Fixed. Ready for review again.

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs (hunk 1 above)
Comment (hishitetsu) [reply src/Files.App/ViewModels/Properties/HashesViewModel.cs:93]: Done.

File: src/Files.App/Views/Properties/HashesPage.xaml (hunk 2 above)
Comment (hishitetsu) [reply src/Files.App/Views/Properties/HashesPage.xaml:104]: > Use Converter 'BoolToVisibility' since IsEnabled must be boolean.

I thought so too, but it actually works as it is.

> Remove Mode=OneWay due to a warning WMC1506.

No. The source property change is not detected without `Mode=OneWay`

File: src/Files.App/Views/Properties/HashesPage.xaml (hunk 5 above)
Comment (hishitetsu) [reply src/Files.App/Views/Properties/HashesPage.xaml]: I decided to read the file for each algorithm, so I removed this text block.

File: src/Files.App/Views/Properties/HashesPage.xaml.cs (hunk 6 above)
Comment (hishitetsu) [reply src/Files.App/Views/Properties/HashesPage.xaml.cs]: Done.

File: src/Files.App/Views/Properties/HashesPage.xaml.cs (hunk 7 above)
Comment (hishitetsu) [reply src/Files.App/Views/Properties/HashesPage.xaml.cs]: Done.

File: src/Files.App/Views/Settings/AboutPage.xaml (hunk 8 above)
Comment (hishitetsu) [reply src/Files.App/Views/Settings/AboutPage.xaml]: Reverted.

> please move methods you've created for 3rd-party stuff in AboutPage's code-behind to the ViewModel.

Which methods?

**Comment by yaira2** [comment]
@onein528 do you want to review this?

**Comment by onein528** [comment]
Yes. @hishitetsu is doing pretty nice work.

File: src/Files.App/Views/Properties/HashesPage.xaml [hunk 9]
```diff
@@ -22,25 +21,10 @@
...
-			VerticalAlignment="Center"
-			x:Load="{x:Bind HashesViewModel.IsLoading, Mode=OneWay}"
-			Spacing="8">
-			<ProgressRing HorizontalAlignment="Center" IsIndeterminate="True" />
-			<TextBlock
-				x:Name="LoadingTextBlock"
-				HorizontalAlignment="Center"
-				Style="{StaticResource BodyStrongTextBlockStyle}"
-				Text="{helpers:ResourceString Name=Loading}" />
-		</StackPanel>
```
Comment (onein528) [inline src/Files.App/Views/Properties/HashesPage.xaml:37]: Why removed?

**Review by onein528 — COMMENTED** [review]
I don't have my PC right now, but nice! UI changes should be reviewed by somebody.

File: src/Files.App/Views/Properties/HashesPage.xaml (hunk 9 above)
Comment (hishitetsu) [reply src/Files.App/Views/Properties/HashesPage.xaml:37]: Loading the file and calclulating the hash is now processed when `isEnabled` property is changed, not the page is loaded.
So I remove this because there are no possibility to take a long time to load the page.

**Comment by hishitetsu** [comment]
Fixed the bug and Refactored. Now ready for review again.

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 10]
```diff
@@ -41,76 +28,89 @@ public HashInfoItem SelectedItem
...
 			{
-				CanAccessFile = false;
-			}
-			finally
+				new() { Algorithm = "MD5" },
+				new() { Algorithm = "SHA1" },
+				new() { Algorithm = "SHA256" },
+				new() { Algorithm = "SHA384" },
+				new() { Algorithm = "SHA512" },
+			};
```
Comment (onein528) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:50]: Add a line between.

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 11]
```diff
@@ -41,76 +28,89 @@ public HashInfoItem SelectedItem
...
+				else
+				{
+					x.IsEnabled = x.Algorithm switch
+					{
+						"MD5" => true,
+						"SHA1" => true,
+						"SHA256" => true,
+						"SHA384" => false,
+						"SHA512" => false,
+						_ => throw new InvalidOperationException()
```
Comment (onein528) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:66]: Should not occur an exception. Please set default like MD5

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 12]
```diff
@@ -41,76 +28,89 @@ public HashInfoItem SelectedItem
...
-				Algorithm = "SHA384",
-				HashValue = await ChecksumHelpers.CreateSHA384(_stream, cancellationToken),
-			});
-			Hashes.Add(new()
-			{
-				Algorithm = "SHA512",
-				HashValue = await ChecksumHelpers.CreateSHA512(_stream, cancellationToken),
-			});
+				if (!_showHashesDictionary.ContainsKey(hashInfoItem.Algorithm)
+					|| _showHashesDictionary[hashInfoItem.Algorithm] != hashInfoItem.IsEnabled)
```
Comment (onein528) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:77]:
```suggestion
				if (!_showHashesDictionary.ContainsKey(hashInfoItem.Algorithm) ||
					_showHashesDictionary[hashInfoItem.Algorithm] != hashInfoItem.IsEnabled)
```

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 13]
```diff
@@ -41,76 +28,89 @@ public HashInfoItem SelectedItem
...
+				{
+					_showHashesDictionary[hashInfoItem.Algorithm] = hashInfoItem.IsEnabled;
+					UserSettingsService.PreferencesSettingsService.ShowHashesDictionary = _showHashesDictionary;
+				}
+
+				if (hashInfoItem.HashValue is null && hashInfoItem.IsEnabled)
+				{
+					hashInfoItem.HashValue = "Calculating".GetLocalizedResource();
```
Comment (onein528) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:85]: Add a line between.

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs [hunk 14]
```diff
@@ -41,76 +28,89 @@ public HashInfoItem SelectedItem
...
+							{
+								"MD5" => await ChecksumHelpers.CreateMD5(stream, _cancellationTokenSource.Token),
+								"SHA1" => await ChecksumHelpers.CreateSHA1(stream, _cancellationTokenSource.Token),
+								"SHA256" => await ChecksumHelpers.CreateSHA256(stream, _cancellationTokenSource.Token),
+								"SHA384" => await ChecksumHelpers.CreateSHA384(stream, _cancellationTokenSource.Token),
+								"SHA512" => await ChecksumHelpers.CreateSHA512(stream, _cancellationTokenSource.Token),
+								_ => throw new InvalidOperationException()
+							};
+						}
+						hashInfoItem.IsCalculated = true;
```
Comment (onein528) [inline src/Files.App/ViewModels/Properties/HashesViewModel.cs:100]: Add a line between.

**Review by onein528 — COMMENTED** [review]
Using array in toggle buttons is not good. If you have been able to change, please use command and command parameter.

File: src/Files.App/ViewModels/Properties/HashesViewModel.cs (hunk 11 above)
Comment (hishitetsu) [reply src/Files.App/ViewModels/Properties/HashesViewModel.cs:66]: The default value should be explicitly specified for each algorithm. So I think it should be leave as it is in case someone forgets to set the default value when adding an algorithm.

Comment (onein528) [reply src/Files.App/ViewModels/Properties/HashesViewModel.cs:66]: Should not break the app.

**Comment by hishitetsu** [comment]
@onein528 Updated to use Command. How about this?

**Review by yaira2 — APPROVED** [review]
LGTM
